### DISCUSSION TITLE
Bug Report with LVGate

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#189] [Internal Design] [answered]
Hi Hantao,

There is an initialization bug with ``LVGate``. I was developing ``ESAC1A``, and the code was in my test branch [exctest](https://github.com/jinningwang/andes/tree/exctest).

Below are the diagram and the code snippet.
![image](https://user-images.githubusercontent.com/79226045/135361717-8f3678dd-d406-4f76-b5c3-3e5ce4c72baf.png)

```python
        self.HVG = HVGate(u1=self.UEL,
                          u2=self.LA_y,
                          info='HVGate for under excitation',
                          )

        self.LVG = LVGate(u1=self.HVG_y,
                          u2=self.OEL,
                          info='LVGate for under excitation',
                          )
```
The ``LVGate`` should output the lower value as design. I set ``UEL=-999`` and ``OEL=999``, so the final output should be the input value ``LA_y``. However, the actual output is 0. It was wired.
![image](https://user-images.githubusercontent.com/79226045/135362149-f14767b8-4bee-4307-aeff-6f4c5b200ca2.png)

Could you please check the related function?

Regards,
Jinning

**Comment by cuihantao** [comment]
Thanks for reporting. This is a bug.

In LVGate, we need to make sure that the Selector is evaluated before updating the equation. Inside the definition of LVGate, move the selector to before the variable, the provide discrete=self.sel to the variable. See if this helps.

If not, we need to look at the input to the initialization equation.

If you put a breakpoint in the initialization equation for LVG_y in ~/.andes/pycode/ESAC1A, you can check the values of the arguments used for evaluating that equation for initial value.

I suspect that the values input to that initialization function is different than the final output. That is, there is a sequence issue.

**Comment by jinningwang** [comment]
Hi Hantao,

I modified the ``LVGate`` definitions as below:
```python
        ...
        self.sl = Selector(self.u1, self.u2, fun=np.minimum.reduce,
                           info='LVGate Selector',
                           )
        self.y = Algeb(info='LVGate output', tex_name='y', discrete=self.sl)

        self.vars = {'sl': self.sl, 'y': self.y}
        ...
```
However, the bug was still there.

Then I checked the pycode, the ``LVG_y`` shows:
```python
def LVG_y_ia(HVG_y, OEL, LVG_sl_s1, LVG_sl_s0):
    return (HVG_y*LVG_sl_s0 + LVG_sl_s1*OEL)
```

The sequence seems correct, but the bug remains there.

**Reply by cuihantao** [reply]
Which branch has your latest code? 

It fixes the issue after the modification you put.

**Reply by cuihantao** [reply]
I just pushed my changes to my `develop`. You can merge it with yours - if you have committed after modifying `block.py`, you will need manually resolve the conflict.

**Comment by cuihantao** [comment]
I placed a breakpoint at `    return (HVG_y*LVG_sl_s0 + LVG_sl_s1*OEL)` to inspect the values.

It turns out that both `LVG_sl_s0` and `LVG_sl_s1` are zeros. Let me keep digging.

**Reply by jinningwang** [reply]
Sorry, I haven't pushed my code yet. Got a meeting at that time.

Thanks!

Regards,
Jinning

**Comment by jinningwang** [comment, marked as answer]
The bug was fixed by modifying the ``LVGate`` definition as below:
```
        ...
        self.sl = Selector(self.u1, self.u2, fun=np.minimum.reduce,
                           info='LVGate Selector',
                           )
        self.y = Algeb(info='LVGate output', tex_name='y', discrete=self.sl)

        self.vars = {'sl': self.sl, 'y': self.y}
        ...
```

Thanks for Hantao's help!
